delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/eventlog/EventExchange.scala

```diff
@@ -0,0 +1,49 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.eventlog
+
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.{CompactedJsonLd, ExpandedJsonLd}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{Event, ResourceF}
+import com.typesafe.scalalogging.Logger
+import monix.bio.Task
+
+import scala.reflect.ClassTag
+
+/**
+  * Resolver which allows to fetch the latest state for an [[Event]].
+  */
+trait EventExchange[E <: Event] {
```

Comment:
This way you decided it to do it, instead of the one I provided as an example has some drawbacks:
1. You have to cast twice (one when calling appliesTo and another when calling toExpanded or toCompacted)
3. You have to provide many more methods on the same class.
3. You have to define the generic method `toCompacted` and `toExpanded` implementing this logger which usually it should't be required at all, since the logging already happens on the GlobalEventLog
3. You have this generic type which gets passed along everywhere (event though it is useless cause it ends up being `_`), instead of an abstract type which does not pollute the rest of the code.

Having an intermediate class `EventExchangeApply` (or whatever you wanna call it) solves most of the issues and keeps the class `EventExchange` with a single responsibility.

Reply:
If we have something like this:
```scala
EventExchange[E <: Event, S](fetch: E => IO[SomeError, State])(implicit c: ClassTag, jsonldEncoder: JsonLdEncoder[S])
```
Wouldn't that make things simpler and remove the need of a trait as we would have all what we need ?

Reply:
I don't see the signature to that `EventExchange` more useful than the current. The `JsonLdEncoder[S]` part is misleading, since in many cases you don't want to perform any encoding (it is already present in the State).

Reply:
I should not have put state but as it is the resource built from the state, my bad.
And we don't know at this level that it works that way and we should not care, it is up to the JsonLdEncoder that we already use for the different types of resources to take care of it.

Reply:
@umbreak 
re 1. the cast in `appliesTo` is only done once, when building the index, 
re2.  I find it cleaner to have those methods in a single class rather than spread through two
re3 logging in 2 places notifies about two possible errors: no `EventExchange` for a given event and wrong event passed to event exchange(the latter shouldn't happen though).
re4 I agree that having `[_ <: Event]` everywhere is quite annoying, but I personally find the version with  `EventExchangeApply` quite difficult to follow.

Reply:
I'm fine if you wanna keep this version. The please try the following:
- Try to use an abstract type instead of the generic type. It should work and you should be able to get rid of the `[_ <: Event]`
- Remove the logging from here. This is a log level implementation, which returns a Task[Option] which is just fine. The logging, if necessary, should be done in the higher level place where this is used. And I believe is not even necessary, since this is already handled in the globaleventlog